python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1380 @@
+        dtype=float,
+        doc="Number of sigma to clip to during pixel-wise clipping when generating the kernel. See "
+        "the generateKernel docstring for more info.",
+        default=4
+    )
+    nSigmaClipXCorr = pexConfig.Field(
+        dtype=float,
+        doc="Number of sigma to clip when calculating means for the cross-correlation",
+        default=5
+    )
+    maxLag = pexConfig.Field(
+        dtype=int,
+        doc="The maximum lag (in pixels) to use when calculating the cross-correlation/kernel",
+        default=8
+    )
+    nPixBorderGainCalc = pexConfig.Field(
+        dtype=int,
+        doc="The number of border pixels to exclude when calculating the gain",
+        default=10
+    )
+    nPixBorderXCorr = pexConfig.Field(
+        dtype=int,
+        doc="The number of border pixels to exclude when calculating the cross-correlation and kernel",
+        default=10
+    )
+    biasCorr = pexConfig.Field(
+        dtype=float,
+        doc="An empirically determined correction factor, used to correct for the sigma-clipping of" +
+        " a non-Gaussian distribution. Post DM-15277, code will exist here to calulate appropriate values",
+        default=0.9241
+    )
+    backgroundBinSize = pexConfig.Field(
+        dtype=int,
+        doc="Size of the background bins",
+        default=128
+    )
+    fixPtcThroughOrigin = pexConfig.Field(
+        dtype=bool,
+        doc="Contrain the fit of the PTC to go through the origin?",
+        default=True
+    )
+    level = pexConfig.ChoiceField(
+        doc="The level at which to calculate kernels",
+        dtype=str, default="DETECTOR",
+        allowed={
+            "AMP": "Every amplifier treated separately",
+            "DETECTOR": "One kernel per detector",
+        }
+    )
+    backgroundWarnLevel = pexConfig.Field(
+        dtype=float,
+        doc="Log warnings if the mean of the fitted background is found to be above this level after " +
+        "differencing image pair.",
+        default=0.1
+    )
+
+
+class MakeBrighterFatterKernelTaskRunner(pipeBase.TaskRunner):
+    """Subclass of TaskRunner for the MakeBrighterFatterKernelTask.
+
+    This transforms the processed arguments generated by the ArgumentParser
+    into the arguments expected by makeBrighterFatterKernelTask.run().
+
+    makeBrighterFatterKernelTask.run() takes a two arguments,
+    one of which is the dataRef (as usual), and the other is the list
+    of visitPairs, in the form of a list of tuples.
+    This list is supplied on the command line as documented,
+    and this class parses that, and passes the parsed version
+    to the run() method.
+
+    See pipeBase.TaskRunner for more information.
+    """
+
+    @staticmethod
+    def getTargetList(parsedCmd, **kwargs):
+        """Parse the visit list and pass through explicitly."""
+        visitPairs = [(int(visit.split(",")[0]), int(visit.split(",")[1])) for visit in parsedCmd.visitPairs]
```

Comment:
I strongly suggest better error checking and reporting.

Heres is a version with good error checking that uses the callback mechanism. I **have** tested this code and hope you will consider using it. It parses the data in the usual location, and thus produces really nice error messages that don't include distracting tracebacks.
```
class VisitPairAction(argparse.Action):
    """argparse action callback to process a visit pair into
    a tuple 2 floats
    """
    def __call__(self, parser, namespace, values, option_string):
        """Parse ``--visitPairs`` data and append results to
        ``namespace.<argument>.visitPairs``.

        Parameters
        ----------
        parser : `ArgumentParser`
            Argument parser.
        namespace : `argparse.Namespace`
            Parsed command (an instance of argparse.Namespace).
            The following attributes are updated:

            - ``visitPairs``
        values : `list`
            A list of visit pairs, each of the form "visit,visit"
        option_string : `str`
            Option value specified by the user. Ignored.
        """
        if namespace.visitPairs is None:
            namespace.visitPairs = []
        for visitStringPair in values:
            visitStrings = visitStringPair.split(",")
            if len(visitStrings) != 2:
                parser.error("Found {} visits in {} instead of 2".format(len(visitStrings), visitStringPair))
            try:
                visits = [int(visit) for visit in visitStrings]
            except Exception:
                parser.error("Could not parse {} as two integer visit numbers".format(visitStringPair))
            namespace.visitPairs.append(visits)
```
If you are really keen to parse the data in the task runner then you can easily adapt the above, but error messages will produce tracebacks and will not print help for the task.

Doing it in a loop instead of a list comprehension is slower and more verbose, but speed should not be an issue and good error reporting is important. For instance users will occasionally forget and put a space after a comma: "--visit-pairs 1, 2 3, 4" and a good error message is helpful. You could also accept that, but it would require more code -- especially when using the callback mechanism, as you'd have to keep track of "did the last value end with a comma, if so the next value should be a single integer". I don't think it is necessary or expected.

Reply:
Thanks for this, and yes, I had indeed just missed it before. I've split the difference and implemented what you put here (and many thanks for that!), but not in the argparser as this avoids callbacks which I think aren't strictly necessary and would make running the task more magic than it needs to be.